### PR TITLE
Support vote delegation

### PR DESCRIPTION
**Pull request by Orelsanpls**
Vote / Withdraw / Sync using a single transaction

### CODE REVIEW

**Comment by acamill** [comment]
I do have staked UXP on the stake website but I don't see them on the governance side I think it's normal since I didn't do the conversion, how would this play out in a real situation?

**Comment by Orelsanpls** [comment]
> I do have staked UXP on the stake website but I don't see them on the governance side I think it's normal since I didn't do the conversion, how would this play out in a real situation?

In the real situation, you would see :

Screenshot 2022-05-31 at 16 59 39

Then 

Screenshot 2022-05-31 at 17 00 47

After voting.

For now, we do no adapt the display about the amount of tokens underneath the delegate accounts

File: actions/castVotes.ts [hunk 1]
```diff
@@ -23,15 +19,23 @@ import { sendTransaction } from '../utils/send';
 import { withUpdateVoterWeightRecord } from 'VoteStakeRegistry/sdk/withUpdateVoterWeightRecord';
 import { VsrClient } from '@blockworks-foundation/voter-stake-registry-client';
```
Comment (0xCryptoSheik) [inline actions/castVotes.ts]: re-organize imports?

File: components/VotePanel.tsx [hunk 2]
```diff
@@ -12,77 +16,317 @@ import { GoverningTokenType } from '@solana/spl-governance';
...
   const connection = useWalletStore((s) => s.connection);
-  const connected = useWalletStore((s) => s.connected);
+  const fetchVoteRecords = useWalletStore((s) => s.actions.fetchVoteRecords);
   const fetchRealm = useWalletStore((s) => s.actions.fetchRealm);
   const hasVoteTimeExpired = useHasVoteTimeExpired(governance, proposal!);
 
-  const ownVoteRecord =
-    wallet?.publicKey && voteRecordsByVoter[wallet.publicKey.toBase58()];
+  // Look for accounts where the user is the delegate
+  const getDelegatedAccounts = useCallback((): {
```
Comment (0xCryptoSheik) [inline components/VotePanel.tsx]: Unnecessary useCallback

File: components/VotePanel.tsx [hunk 3]
```diff
@@ -12,77 +16,317 @@ import { GoverningTokenType } from '@solana/spl-governance';
...
-  const isVoting =
-    proposal?.account.state === EnhancedProposalState.Voting &&
-    !hasVoteTimeExpired;
+  const handleShowVoteModal = (vote: YesNoVote) => {
+    setVote(vote);
+
+    setShowVoteModal(true);
+  };
+
+  const handleCloseShowVoteModal = useCallback(() => {
```
Comment (0xCryptoSheik) [inline components/VotePanel.tsx]: unnecessary useCallback

File: components/VotePanel.tsx [hunk 4]
```diff
@@ -12,77 +16,317 @@ import { GoverningTokenType } from '@solana/spl-governance';
...
+    setShowVoteModal(false);
+  }, []);
+
+  // Show nothing if not connected
+  if (!wallet?.publicKey) {
+    return <></>;
```
Comment (0xCryptoSheik) [inline components/VotePanel.tsx]: `return null;` ?

File: components/TokenBalance/TokenBalanceCard.tsx [hunk 5]
```diff
@@ -106,12 +137,66 @@ const TokenDeposit = ({
...
+    }
+
+    return Object.entries(tokenRecords)
+      .filter(([, value]) =>
+        value.account.governanceDelegate?.equals(wallet.publicKey!),
+      )
+      .map(([key, value]) => ({
+        address: key,
+        nbToken: value.account.governingTokenDepositAmount
+          .div(new BN(10 ** 6))
```
Comment (0xCryptoSheik) [inline components/TokenBalance/TokenBalanceCard.tsx]: This is supposed to be UXP? UXP has 9 decimals?

File: components/VotePanel.tsx [hunk 6]
```diff
@@ -12,77 +16,302 @@ import { GoverningTokenType } from '@solana/spl-governance';
...
+  return accountsToVoteFor.reduce(
+    (acc, account) => {
+      // Ignore account without voterTokenRecord
+      if (!account.voterTokenRecord) {
+        return acc;
+      }
+
+      if (
+        account.voterTokenRecord.account.governingTokenDepositAmount.eq(
+          new BN(0),
```
Comment (0xCryptoSheik) [inline components/VotePanel.tsx]: nit: declare in a variable outside of the component `BN_ZERO`

File: components/VotePanel.tsx [hunk 7]
```diff
@@ -12,77 +16,302 @@ import { GoverningTokenType } from '@solana/spl-governance';
...
-      ? ownTokenRecord
-      : ownCouncilTokenRecord;
+    return Object.entries(tokenRecords)
+      .filter(([, value]) =>
+        value.account.governanceDelegate?.equals(wallet.publicKey!),
+      )
+      .map(([key, value]) => ({
+        address: key,
+        nbToken: value.account.governingTokenDepositAmount
+          .div(new BN(10 ** 6))
```
Comment (0xCryptoSheik) [inline components/VotePanel.tsx]: Consider using a helper, and take into account mint decimals

File: hooks/useHotWalletPluginUXDStaking.ts [hunk 8]
```diff
@@ -60,48 +65,54 @@ const useHotWalletPluginUXDStaking = (hotWalletAccount: HotWalletAccount) => {
         );
       }
 
+      const sssClient = SingleSideStakingClient.load({
+        provider: new SolanaAugmentedProvider(
```
Comment (0xCryptoSheik) [inline hooks/useHotWalletPluginUXDStaking.ts]: Didn't we remove that from the consumer & hide it in the client instead?

File: pages/dao/[symbol]/proposal/components/instructions/UXDStaking/InitializeStakingCampaign.tsx [hunk 9]
```diff
@@ -67,9 +69,19 @@ const InitializeStakingCampaign = ({
         );
       }
 
-      const client: SingleSideStakingClient = new SingleSideStakingClient(
+      if (!realmInfo) {
+        throw new Error('Realm info not loaded');
+      }
+
+      const sssClient = SingleSideStakingClient.load({
+        provider: new SolanaAugmentedProvider(
```
Comment (0xCryptoSheik) [inline pages/dao/[symbol]/proposal/components/instructions/UXDStaking/InitializeStakingCampaign.tsx]: Same, also shouldn't we centralize staking client initialization somewhere ?

File: pages/dao/[symbol]/proposal/components/instructions/UXDStaking/MigrateStakingCampaignFromV1ToV2.tsx [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+          );
+        }
+
+        if (!realmInfo) {
+          throw new Error('Realm info not loaded');
+        }
+
+        const stakingCampaignPda = new PublicKey(form.stakingCampaignPda!);
+
+        const sssClient = SingleSideStakingClient.load({
```
Comment (0xCryptoSheik) [inline pages/dao/[symbol]/proposal/components/instructions/UXDStaking/MigrateStakingCampaignFromV1ToV2.tsx]: again

File: hooks/useHotWalletPluginUXDStaking.ts (hunk 8 above)
Comment (Orelsanpls) [reply hooks/useHotWalletPluginUXDStaking.ts]: Yes the governance have not updated to the latest client yet, will do

File: components/TokenBalance/TokenBalanceCard.tsx (hunk 5 above)
Comment (Orelsanpls) [reply components/TokenBalance/TokenBalanceCard.tsx]: How did I let this slip :o 

really good catch. Never say 'I'll see about that later' as you will 100% forget, make it clean right away

File: components/VotePanel.tsx (hunk 6 above)
Comment (Orelsanpls) [reply components/VotePanel.tsx]: Screenshot 2022-06-17 at 11 45 58

Comment (Orelsanpls) [reply components/VotePanel.tsx]: indeed

File: hooks/useHotWalletPluginUXDStaking.ts (hunk 8 above)
Comment (Orelsanpls) [reply hooks/useHotWalletPluginUXDStaking.ts]: Done :)

File: pages/dao/[symbol]/proposal/components/instructions/UXDStaking/InitializeStakingCampaign.tsx (hunk 9 above)
Comment (Orelsanpls) [reply pages/dao/[symbol]/proposal/components/instructions/UXDStaking/InitializeStakingCampaign.tsx]: Yes, good idea, time to cleanup a bit :)

File: components/TokenBalance/TokenBalanceCard.tsx (hunk 5 above)
Comment (acamill) [reply components/TokenBalance/TokenBalanceCard.tsx]: Nice catch Sheik 🤭

**Pull request merged**